barretenberg/cpp/src/barretenberg/dsl/acir_format/proof_surgeon.hpp

```diff
@@ -0,0 +1,116 @@
+    {
+        std::vector<uint32_t> proof;
+        proof.reserve(proof_in.size() + public_inputs.size());
+
+        // Construct a the complete proof as the concatenation {"initial data" | public_inputs | proof_in}
+        proof.insert(proof.end(), proof_in.begin(), proof_in.begin() + HONK_RECURSION_PUBLIC_INPUT_OFFSET);
+        proof.insert(proof.end(), public_inputs.begin(), public_inputs.end());
+        proof.insert(proof.end(), proof_in.begin() + HONK_RECURSION_PUBLIC_INPUT_OFFSET, proof_in.end());
+
+        return proof;
+    }
+
+    /**
+     * @brief Extract then remove a given number of public inputs from a proof
+     *
+     * @param proof_witnesses Witness values of a bberg style proof containing public inputs
+     * @param num_public_inputs The number of public inputs to extract from the proof
+     * @return std::vector<bb::fr> The extracted public input witness values
+     */
+    static std::vector<bb::fr> cut_public_inputs_from_proof(std::vector<bb::fr>& proof_witnesses,
+                                                            const size_t num_public_inputs_to_extract)
+    {
+        // Construct iterators pointing to the start and end of the public inputs within the proof
+        auto pub_inputs_begin_itr =
+            proof_witnesses.begin() + static_cast<std::ptrdiff_t>(HONK_RECURSION_PUBLIC_INPUT_OFFSET);
+        auto pub_inputs_end_itr =
+            proof_witnesses.begin() +
+            static_cast<std::ptrdiff_t>(HONK_RECURSION_PUBLIC_INPUT_OFFSET + num_public_inputs_to_extract);
+
+        // Construct the isolated public inputs
+        std::vector<bb::fr> public_input_witnesses{ pub_inputs_begin_itr, pub_inputs_end_itr };
+
+        // Erase the public inputs from the proof
+        proof_witnesses.erase(pub_inputs_begin_itr, pub_inputs_end_itr);
+
+        return public_input_witnesses;
+    }
+
+    struct RecursionWitnessData {
+        std::vector<uint32_t> key_indices;
+        std::vector<uint32_t> proof_indices;
+        std::vector<uint32_t> public_inputs_indices;
+    };
+
+    /**
+     * @brief Populate a witness vector with key, proof, and public inputs; track witness indices for each component
+     * @details This method is used to constuct acir-style inputs to a recursion constraint. It is assumed that the
+     * provided proof contains all of its public inputs (i.e. the conventional bberg format) which are extracted herein.
+     * Each component is appended to the witness which may already contain data. The order in which they are added is
+     * arbitrary as long as the corresponding witness indices are correct.
+     *
+     * @param witness
+     * @param proof_witnesses
+     * @param key_witnesses
+     * @param num_public_inputs
+     * @return RecursionWitnessData
+     */
+    static RecursionWitnessData populate_recursion_witness_data(bb::SlabVector<bb::fr>& witness,
+                                                                std::vector<bb::fr>& proof_witnesses,
```

Comment:
This is a pretty substantial simplification of the old logic, driven in part by the fact that there's no reason (as far as I can tell) to place the data into `witness` in any special way, as long as the component indices are correct. So I just basically concatenate the key, proof and pub inputs into with witness without worrying about reinserting the pub inputs in the proof etc.